hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AbstractPartitionMessageTask.java

```diff
@@ -57,17 +61,28 @@ public final void processMessage() {
         beforeProcess();
         Operation op = prepareOperation();
         op.setCallerUuid(endpoint.getUuid());
-        InvocationBuilder builder = nodeEngine.getOperationService()
+        ICompletableFuture f = nodeEngine.getOperationService()
                 .createInvocationBuilder(getServiceName(), op, getPartitionId())
-                .setExecutionCallback(this)
-                .setResultDeserialized(false);
+                .setResultDeserialized(false)
+                .invoke();
 
-        builder.invoke();
+        f.andThen(this, this);
     }
 
     protected abstract Operation prepareOperation();
 
     @Override
+    public void execute(Runnable command) {
+        if (Thread.currentThread() instanceof OperationHostileThread) {
```

Comment:
isn't it better to change the logic? 

if the current thread is `PartitionThread` then use it directly. otherwise offload to the ASYNC_EXECUTOR. It makes it more explicit when the current thread is used.